### PR TITLE
Add feature to configure IMUs at initialization

### PR DESCRIPTION
**Pull request by CatYang3**
# Summary
Describe the main changes in this PR:
- Added feature to configure IMUs at init stage, including setting of ranges and data rates, enabling features (for the BNO08x), etc. by defining the steps in IMUConfig
- Removed the IMU frequency settings from the sensor manager, so that it checks for new data at a rate faster than the IMU frequency
- In order for the sensor manager to check for new data, IMUData assembly is moved from the Wrapper to the sensor manager. The Wrapper only returns a new dataclass called IMURawData (accel, gyro, mag).
- Done: The BNO055 has to switch to a config mode before the ranges and frequencies can be changed. It will fail otherwise
  - Because the mode switch should be followed by time.sleep(), the pre-config logic is updated to support python libraries like ```time```.
- TODO: Have the logger make warnings when readings get clipped (in another PR in the future).
---

# Code Quality Checklist
Before requesting review, ensure:
- [x] All CI/CD pipelines pass
- [x] Code includes docstrings and type hints
- [x] I added new tests where necessary
- [x] I ran **`make tree`** to update the README.md
---

# Additional Notes
- I realize this PR has a lot of changes. The code was written with compatibility in mind, so it works with at least the 3 IMU models we have. I also tried to verify the code with tests that cover as much as possible.

### CODE REVIEW

**Comment by CatYang3** [comment]
I just found out that the BNO055 might also need extra steps before I could set its ranges, so if the code is run in its current state with the BNO055 plugged in will not work: ```RuntimeError: Mode must not be a fusion mode```

File: src/imu_python/sensor_manager.py [hunk 1]
```diff
@@ -61,11 +59,23 @@ def _loop(self) -> None:
             try:
                 # Attempt to read all sensor data
                 data = self.imu_wrapper.get_data()
-                with self.lock:
-                    self.latest_data = data
-                    if self.log_data:
-                        self.IMUData_log.append(data)
-                time.sleep(self.period)
+                # Ensure new data
+                if self.latest_data is None or data != self.latest_data.raw_data:
```
Comment (Tsmorz) [inline src/imu_python/sensor_manager.py]: Can you add a logger.debug or logger.trace here?

File: src/imu_python/wrapper.py [hunk 2]
```diff
@@ -60,41 +62,125 @@ def read_sensor(self, attr: str) -> VectorXYZ:
...
+                value = getattr(value, part)
+            except AttributeError as err:
+                raise RuntimeError(
+                    f"Failed to resolve attribute '{part}' in module '{module.__name__}'"
+                ) from err
+
+        return value
+
+    def _resolve_func(self, func_name: str) -> Callable:
+        """Resolve a function name to a callable."""
```
Comment (Tsmorz) [inline src/imu_python/wrapper.py]: Needs proper docstring.

File: src/imu_python/wrapper.py [hunk 3]
```diff
@@ -60,41 +62,125 @@ def read_sensor(self, attr: str) -> VectorXYZ:
...
+        except AttributeError as err:
+            raise RuntimeError(
+                f"Failed to resolve attribute '{part}' in module '{root}'"
+            ) from err
+
+        if not callable(func):
+            raise RuntimeError(
+                f"Failed to resolve function '{func_name}' in module '{module}'"
+            )
+        return func
```
Comment (Tsmorz) [inline src/imu_python/wrapper.py]: It's a little unclear what this method does from just reading it as well.

File: src/imu_python/wrapper.py [hunk 4]
```diff
@@ -60,41 +62,125 @@ def read_sensor(self, attr: str) -> VectorXYZ:
...
+                            f"'{step.name}' in module '{self.config.name}' is not a callable"
+                        )
+                except AttributeError:
+                    # try to resolve function globally
+                    func = self._resolve_func(step.name)
+                # call the function with resolved args and kwargs
+                func(*resolved_args, **resolved_kwargs)
+                continue
+
+            elif step.step_type == "set":
```
Comment (Tsmorz) [inline src/imu_python/wrapper.py]: Can you make an Enum class for call and set?

File: src/imu_python/wrapper.py [hunk 5]
```diff
@@ -60,41 +62,125 @@ def read_sensor(self, attr: str) -> VectorXYZ:
...
+        """Dynamically import a Python module by path."""
         try:
-            module = importlib.import_module(self.config.library)
-            return module
+            return importlib.import_module(module_path)
         except ImportError as err:
             raise RuntimeError(
-                f"{err} - Failed to import IMU driver '{self.config.library}'."
+                f"{err} - Failed to import module '{module_path}'."
             ) from err
```
[the remark this answers is not in the document]
Comment (CatYang3) [reply src/imu_python/wrapper.py]: Yes, I've combined them now. The only place the _import_imu_module was used was in the reload() function, so changing the function call there should be all I need to do.

File: src/imu_python/wrapper.py [hunk 6]
```diff
@@ -60,41 +62,130 @@ def read_sensor(self, attr: str) -> VectorXYZ:
         else:
             return VectorXYZ.from_tuple(data)
 
-    def get_data(self) -> IMUData:
+    def get_data(self) -> IMURawData:
```
Comment (Tsmorz) [inline src/imu_python/wrapper.py]: Can you change to get_raw_data()?

File: src/imu_python/wrapper.py [hunk 7]
```diff
@@ -60,41 +62,130 @@ def read_sensor(self, attr: str) -> VectorXYZ:
...
+            value = module
+            for part in rest:
+                value = getattr(value, part)
+            func = value
+        except RuntimeError as err:
+            raise RuntimeError(f"Failed to import '{root}'") from err
+        except AttributeError as err:
+            raise RuntimeError(
+                f"Failed to resolve attribute '{part}' in module '{root}'"
+            ) from err
```
Comment (Tsmorz) [inline src/imu_python/wrapper.py]: Why is this exception changed to raise a RuntimeError when it was an AttributeError initially?

Comment (CatYang3) [reply src/imu_python/wrapper.py]: So that the edge case can be tested together with the other ones with pytest parametrize. Should I change it back and also modify the test as well?
